File: Linked List/leetcode_02.py
# You are given two non - empty linked lists representing two non - negative integers.The digits are stored in reverse order,
# and each of their nodes contains a single digit.Add the two numbers and return the sum as a linked list.
# You may assume the two numbers do not contain any leading zero,except the number 0 itself.

class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next


class Solution:
  def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
    dummyHead = ListNode(0)
    tail = dummyHead
    carry = 0

    while l1 is not None or l2 is not None or carry != 0:
      digit1 = l1.val if l1 is not None else 0
      digit2 = l2.val if l2 is not None else 0

      sum = digit1 + digit2 + carry
      digit = sum % 10
      carry = sum // 10

      newNode = ListNode(digit)
      tail.next = newNode
      tail = tail.next

      l1 = l1.next if l1 is not None else None
      l2 = l2.next if l2 is not None else None

    result = dummyHead.next
    dummyHead.next = None
    return result

def main():
  l1 = ListNode(2)
  l1.next = ListNode(4)
  l1.next.next = ListNode(3)

  l2 = ListNode(5)
  l2.next = ListNode(6)
  l2.next.next = ListNode(4)

  solution = Solution()
  result = solution.addTwoNumbers(l1, l2)

  while result:
    print(result.val, end=" ")
    result = result.next
  print()

if __name__ == "__main__":
  main()

File: Linked List/leetcode_02.cpp

// Problem Link ====>> https://leetcode.com/problems/add-two-numbers/description/

#include <iostream>
#include <vector>
using namespace std;

struct ListNode
{
public:
    int val;
    ListNode *next;

    ListNode(int val)
    {
        this->val = val;
        this->next = nullptr;
    }
};

class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode *dummyHead = new ListNode(0);
        ListNode *tail = dummyHead;
        int carry = 0;

        while (l1 != nullptr || l2 != nullptr || carry != 0)
        {
            int digit1 = (l1 != nullptr) ? l1->val : 0;
            int digit2 = (l2 != nullptr) ? l2->val : 0;

            int sum = digit1 + digit2 + carry;
            int digit = sum % 10;
            carry = sum / 10;

            ListNode *newNode = new ListNode(digit);
            tail->next = newNode;
            tail = tail->next;

            l1 = (l1 != nullptr) ? l1->next : nullptr;
            l2 = (l2 != nullptr) ? l2->next : nullptr;
        }

        ListNode *result = dummyHead->next;
        delete dummyHead;
        return result;
    }
};

// Helper function to print the linked list
void printLinkedList(ListNode *head)
{
    ListNode *temp = head;
    while (temp != nullptr)
    {
        cout << temp->val << " ";
        temp = temp->next;
    }
    cout << endl;
}

int main()
{
    Solution solution;

    // Creating two linked lists
    ListNode *l1 = new ListNode(2);
    l1->next = new ListNode(4);
    l1->next->next = new ListNode(3);

    ListNode *l2 = new ListNode(5);
    l2->next = new ListNode(6);
    l2->next->next = new ListNode(4);

    cout << "First linked list: ";
    printLinkedList(l1);

    cout << "Second linked list: ";
    printLinkedList(l2);

    // Adding two linked lists
    ListNode *result = solution.addTwoNumbers(l1, l2);

    cout << "Result of addition: ";
    printLinkedList(result);

    // Cleanup: Freeing memory allocated to the linked lists
    ListNode *temp;
    while (l1 != nullptr)
    {
        temp = l1;
        l1 = l1->next;
        delete temp;
    }

    while (l2 != nullptr)
    {
        temp = l2;
        l2 = l2->next;
        delete temp;
    }

    while (result != nullptr)
    {
        temp = result;
        result = result->next;
        delete temp;
    }

    return 0;
}